crates/lang_srv/src/registry.rs

```diff
@@ -1,75 +1,193 @@
+                } else {
+                    debug!(
+                        "Document typechecking failed at version {:?}, not updating last_good_document",
+                        &document.doc_info.version
+                    );
+                    *old_doc = DocumentPair::new(document, old_doc.last_good_document.clone());
                 }
             }
-            DocumentChange::Closed(_url) => {
-                // Do nothing.
+            None => {
+                documents.insert(url.clone(), DocumentPair::new(document.clone(), document));
+            }
+        }
+    }
+
+    pub async fn apply_changes<'a>(&self, analysed_docs: Vec<AnalyzedDocument>, updating_url: Url) {
+        let mut documents = self.documents.lock().await;
+        debug!(
+            "finised doc analysis for doc: {:?}",
+            updating_url.to_string()
+        );
+
+        for document in analysed_docs {
+            let document = Arc::new(document);
+            //Write the newly analysed document into the partial document that any request requiring the latest document will be waiting on
+            Registry::update_document(&mut documents, document, &updating_url);
+        }
+    }
+
+    pub async fn apply_doc_info_changes(&self, url: Url, info: DocInfo) {
+        let mut documents_lock = self.documents.lock().await;
+        let doc = documents_lock.get_mut(&url);
+        match doc {
+            Some(a) => {
+                debug!(
+                    "set the docInfo for {:?} to version:{:?}",
+                    url.as_str(),
+                    info.version
+                );
+                *a = DocumentPair {
+                    info,
+                    last_good_document: a.last_good_document.clone(),
+                    latest_document: OnceLock::new(),
+                };
             }
+            None => debug!("no existing docinfo for {:?} ", url.as_str()),
         }
     }
 
-    fn document_by_url(&mut self, url: &Url) -> Option<&mut AnalyzedDocument> {
-        self.documents.get_mut(url)
+    async fn document_info_by_url(&self, url: &Url) -> Option<DocInfo> {
+        self.documents.lock().await.get(url).map(|a| a.info.clone())
+    }
+
+    ///Tries to get the latest document from analysis.
+    ///Gives up and returns none after 5 seconds.
+    async fn latest_document_by_url(&self, url: &Url) -> Option<Arc<AnalyzedDocument>> {
+        let duration = std::time::Duration::from_secs(5);
+        tokio::time::timeout(duration, async {
+            loop {
+                let docs = self.documents.lock().await;
+                if let Some(a) = docs.get(url) {
+                    if let Some(a) = a.latest_document.get() {
+                        return a.clone();
+                    }
+                }
+                drop(docs);
+                tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+            }
+        })
+        .await
+        .ok()
     }
```

Comment:
Could we still do https://github.com/roc-lang/roc/pull/6134#discussion_r1436683953 here? If nothing else is happening, then we are only blocking on the analysis thread, and re-entering the loop is fine - in the future we can add a condvar to fully yield to the kernel until it's time to wake up. I really want to avoid magic sleep numbers.

Reply:
I don't really want to hit the cpu really hard for no reason, seems like a waste of resources particularly on laptops, battery drain, and such.
This will probably take around 500ms-1s to analyze, that means we will likely check 50-100 times.
We could make it 1ms and check 500-1000 times, but any more than that just sounds silly.
Is there any functional issue with sleeping for some short "magic" amount of time?

Reply:
The reason I want to avoid adding any sleeps (especially non-parameterized ones) is because it’s somewhat arbitrary. Over time you can try to tune it, but that’s a lot of effort, and I don’t think that’s what the project benefits from right now.

Re-entering the loop is not a big deal, I think. The CPU is likely to schedule this thread behind while the analysis is working. Besides, doing a yield_now will yield to the other events in the event loop anyway, and if there’s nothing for the CPU to do, and wake on this thread won’t be a big deal.

I guess I’d just rather trust the runtime and stick with (possibly non-blocking) yields, unless there we have traces to show that these sleep numbers matter. If they do, then we should pick the number that’s the best.

I’m just trying to avoid the case where the analysis becomes fast, and then all of a sudden it finished in 100us for some small incremental change, and now we must keep in mind that there are sleeps in certain parts of the codebase.

Reply:
Okay, that's fair, we are basically just swapping theories at this point. It's awkward to test, and I agree with your idea that it could cause issues in the future so for now i say we just merge this and if it comes up as an slowdown issue we fix it then, or we implement a condvar.